### PR TITLE
Add shellcheck and shfmt

### PR DESCRIPTION
**Pull request by meatballhat**
plus address all shellcheck gripes

### CODE REVIEW

File: gimme [hunk 1]
```diff
@@ -95,21 +96,21 @@ _binary() {
...
-	if [ "${arch}" = 'arm' ] ; then
+	if [ "${arch}" = 'arm' ]; then
 		# attempt "armv6l" vs just "arm" first (since that's what's officially published)
 		urls=(
-			"${GIMME_DOWNLOAD_BASE}/go${version}.${GIMME_OS}-${arch}v6l.tar.gz" # go1.6beta2 & go1.6rc1
-			"${GIMME_DOWNLOAD_BASE}/go${version}.${GIMME_OS}-${arch}6.tar.gz" # go1.6beta1
-			"${urls[@]}"
+			"${GIMME_DOWNLOAD_BASE}/go${version}.${GIMME_OS}-${arch}v6l.tar.gz"
+			"${GIMME_DOWNLOAD_BASE}/go${version}.${GIMME_OS}-${arch}6.tar.gz"
+			"${urls[@]}" # go1.6beta2 & go1.6rc1 # go1.6beta1
```
Comment (meatballhat) [inline gimme]: huh, I guess this was `shfmt` 😸

File: gimme [hunk 2]
```diff
@@ -172,35 +173,37 @@ _setup_bootstrap() {
 	local versions=("1.6" "1.5" "1.4")
 
 	# try existing
-	for v in "${versions[@]}" ; do
-		for candidate in "${GIMME_ENV_PREFIX}/go${v}"*".env" ; do
+	for v in "${versions[@]}"; do
+		for candidate in "${GIMME_ENV_PREFIX}/go${v}"*".env"; do
 			if [ -s "$candidate" ]; then
-				export GOROOT_BOOTSTRAP="$(source "${candidate}" 2>/dev/null && go env GOROOT)"
+				# shellcheck source=/dev/null
```
Comment (meatballhat) [inline gimme:181]: I added this `shellcheck` directive due to the dynamic `source`.

File: gimme [hunk 3]
```diff
@@ -413,57 +417,58 @@ fi
...
+if [[ -n "${2}" ]]; then
 	GIMME_VERSION_PREFIX="${2}"
 fi
 
 case "${GIMME_ARCH}" in
 	x86_64) GIMME_ARCH=amd64 ;;
 	x86) GIMME_ARCH=386 ;;
 	arm64)
-		if [[ "${GIMME_GO_VERSION}" < "1.5" ]]; then
+		if [[ "${GIMME_GO_VERSION}" ]] <"1.5"; then
```
Comment (meatballhat) [inline gimme]: Well this is definitely wrong...

**Review by meatballhat — COMMENTED** [review]
Needs some more tweaks post-shfmt.

File: gimme [hunk 4]
```diff
@@ -209,16 +212,16 @@ _compile() {
 		export CC_FOR_TARGET="${GIMME_CC_FOR_TARGET}"
 
 		local make_log="${1}/make.${GOOS}.${GOARCH}.log"
-		if [ "${GIMME_DEBUG}" > "1" ]; then
-			./make.bash 2>&1 | tee $make_log 1>&2 || return 1
+		if [ "${GIMME_DEBUG}" ] >"1"; then
```
Comment (philpennock) [inline gimme]: This is exposing a bug in the existing code.

`[` does not change parser state, so `>` within `[ ... ]` is a shell-redirection, so you redirected to create a file named `1`

You want `[[ ... > ... ]]` using the lexical-parse-changing doubled-square brackets to keep the string comparison, but I suspect you want `-gt` for numerical comparison (which should work with either singled or doubled square brackets), or just `if (( GIMME_DEBUG > 1 )); then` for arithmetic comparison.

File: gimme [hunk 5]
```diff
@@ -230,48 +233,48 @@ EOF
...
-	if ! _can_compile "${1}" >/dev/null 2>&1 ; then
+	if ! _can_compile "${1}" >/dev/null 2>&1; then
 		# if the compile test fails without GOROOT, then we probably need GOROOT
 		echo 'export GOROOT="'"${1}"'";'
 	else
 		echo 'unset GOROOT;'
 	fi
-	echo 'export PATH="'"${1}/bin"':${PATH}";'
-	if [[ -z "${GIMME_SILENT_ENV}" ]] ; then
+	echo "export PATH=\"${1}/bini\":${PATH}"
```
Comment (meatballhat) [inline gimme]: oh my, not sure what happened here

File: gimme (hunk 1 above)
Comment (mvdan) [reply gimme]: Was this a bug? If so, please file an issue :)

File: gimme (hunk 3 above)
Comment (mvdan) [reply gimme]: It is. https://github.com/mvdan/sh/issues/4 is more urgent than I expected.

File: gimme (hunk 5 above)
Comment (mvdan) [reply gimme]: was this also shfmt?

**Comment by mvdan** [comment]
Apologies for hijacking this pull, but this is uncovering too many interesting issues :)

File: gimme (hunk 5 above)
Comment (meatballhat) [reply gimme]: I'm pretty sure this was my silly fingers 😆

File: gimme (hunk 3 above)
Comment (meatballhat) [reply gimme]: I think @philpennock was right, though, in that the pre-`shfmt` code was always returning `false` as it was attempting to redirect a file named `"1.5"` into `${GIMME_GO_VERSION}`.  `shfmt` ist rad!

File: gimme (hunk 1 above)
Comment (meatballhat) [reply gimme]: I'd say it was a bug in my understanding more than anything I'd call a bug in `shfmt` :heart:

File: gimme (hunk 3 above)
Comment (philpennock) [reply gimme]: For clarity: `[` is an alias for `test`, and with the I/O redirection removed the effective argv for the cmdline to be invoked was then `[ "${GIMME_GO_VERSION}" ]`; `test` without a flag, only a string, is a test for the string being not-null, so just is "true as long as $GIMME_GO_VERSION is defined with a non-empty value".  The falsehood comes from the shell being unable to set up input redirection into the `test` command.  The scenario reduced to `true < /nonexistent`

File: gimme (hunk 1 above)
Comment (mvdan) [reply gimme]: It looks like two comments got collapsed into a single one, which is definitely wrong. Will investigate and fix.

File: gimme (hunk 3 above)
Comment (mvdan) [reply gimme]: I'm confused. This is not `test`, this is bash's extended test. This is a valid comparison, which means "smaller in ASCII alphabetical order". `shfmt` is clearly wrong here.

File: gimme (hunk 1 above)
Comment (mvdan) [reply gimme]: fixed.

File: gimme (hunk 3 above)
Comment (mvdan) [reply gimme]: fixed.

Comment (philpennock) [reply gimme]: I got confused; there were two such re-orderings and one was for `[` (later changed in this PR to `[[`); I thought that they both were `[`, when this one actually started out as `[[`.

Sorry for the confusion.

**Comment by mvdan** [comment]
I just released the first version with binaries, in case that helps with testing or integration with CI:

https://github.com/mvdan/sh/releases/tag/v0.1.0

**Pull request merged**